src/groupeddataframe/grouping.jl

```diff
@@ -129,71 +132,103 @@ end
 Base.names(gd::GroupedDataFrame) = names(gd.parent)
 _names(gd::GroupedDataFrame) = _names(gd.parent)
 
-##############################################################################
-##
-## GroupApplied...
-##    the result of a split-apply operation
-##    TODOs:
-##      - better name?
-##      - ref
-##      - keys, vals
-##      - length
-##      - start, next, done -- should this return (k,v) or just v?
-##      - make it a real associative type? Is there a need to look up key columns?
-##
-##############################################################################
-
-"""
-The result of a `map` operation on a GroupedDataFrame; mainly for use
-with `combine`
-
-Not meant to be constructed directly, see `groupby` abnd
-`combine`. Minimal support is provided for this type. `map` is
-provided for a GroupApplied object.
-
-"""
-struct GroupApplied{T<:AbstractDataFrame}
-    gd::GroupedDataFrame
-    vals::Vector{T}
-
-    function (::Type{GroupApplied})(gd::GroupedDataFrame, vals::Vector)
-        length(gd) == length(vals) ||
-            throw(DimensionMismatch("GroupApplied requires keys and vals be of equal length (got $(length(gd)) and $(length(vals)))."))
-        new{eltype(vals)}(gd, vals)
-    end
-end
-
-
-#
-# Apply / map
-#
-
-# map() sweeps along groups
-function Base.map(f::Function, gd::GroupedDataFrame)
-    GroupApplied(gd, [wrap(f(df)) for df in gd])
-end
-function Base.map(f::Function, ga::GroupApplied)
-    GroupApplied(ga.gd, [wrap(f(df)) for df in ga.vals])
-end
-
 wrap(df::AbstractDataFrame) = df
+wrap(nt::NamedTuple) = nt
+wrap(r::DataFrameRow) = r
 wrap(A::Matrix) = convert(DataFrame, A)
-wrap(s::Any) = DataFrame(x1 = s)
+wrap(s::AbstractVector) = DataFrame(x1 = s)
+wrap(s::Any) = (x1 = s,)
 
 """
-Combine a GroupApplied object (rudimentary)
+    map(f::Function, gd::GroupedDataFrame)
+
+Apply a function to each group of rows and return a `GroupedDataFrame`.
+
+For each group in `gd`, `f` is passed a `SubDataFrame` view holding the corresponding rows.
+`f` can return a single value, a named tuple, a vector, or a data frame.
```

Comment:
can it return a `DataFrameRow` or a `Matrix`?

Reply:
Also below - `Matrix` should be added to the description where we say about a data frame.

Reply:
I've modified the wording a bit to avoid repetition and mention matrices.

Reply:
also please mention `DataFrameRow` (I understand you have not pushed the changes yet as I do not see them).